### PR TITLE
fix http output to never have both transfer-encoding and content-length

### LINKED ISSUE #1765: HTTP API - cat output has both Content-Length and Transfer-Encoding
From [RFC2616#4.4](http://www.w3.org/Protocols/rfc2616/rfc2616-sec4.html#sec4.4):
> Messages MUST NOT include both a Content-Length header field and a non-identity transfer-coding.

The `cat` http response message includes both a content-length header field and a non-identity transfer-encoding.

proof:
```bash
$ echo 'test' > test.txt
$ ipfs add test.txt 
added QmeomffUNfmQy76CQGy9NdmqEnnHU9soCexBnGU3ezPHVH test.txt
$ curl -i http://localhost:5001/api/v0/cat?arg=QmeomffUNfmQy76CQGy9NdmqEnnHU9soCexBnGU3ezPHVH                                            
HTTP/1.1 200 OK
Content-Length: 5
Transfer-Encoding: chunked
X-Stream-Output: 1

test
```
for comparison, the `block get` output:
```bash
$ curl -i http://localhost:5001/api/v0/block/get?arg=QmeomffUNfmQy76CQGy9NdmqEnnHU9soCexBnGU3ezPHVH                                       
HTTP/1.1 200 OK
Transfer-Encoding: chunked
X-Stream-Output: 1



test
```

### PR DESCRIPTION
**Pull request by WeMeetAgain**
Fixes #1765

Since our http responses are unconditionally sending a non-identity `Transfer-Encoding` header, we shouldn't send a `Content-Length` Header.

This patch changes the output `Content-Length` to `X-Content-Length` to resolve that issue and adds a test to confirm that `Transfer-Encoding` and `Content-Length` are not both specified.

### CODE REVIEW

File: test/sharness/t0231-cat-http-no-content-length.sh [hunk 1]
```diff
@@ -0,0 +1,40 @@
...
+		cat cat_output | grep -c Transfer-Encoding >has_te
+	'
+
+	test_expect_success "Search for Content-Length succeeds" '
+		cat cat_output | grep -c Content-Length >has_cl ||
+		echo succeed
+	'
+
+	test_expect_success "Only one of Transfer-Encoding or Content-Length exists" '
+		! [ `cat has_te` -ne 0 -a `cat has_cl` -ne 0 ]
```
Comment (jbenet) [inline test/sharness/t0231-cat-http-no-content-length.sh]: can probably use 
```
! [ `test -f has_te` -a `test -f has_cl` ]
```

File: commands/http/handler.go [hunk 2]
```diff
@@ -231,6 +227,7 @@ func sendResponse(w http.ResponseWriter, r *http.Request, res cmds.Response, req
 	if mime != "" {
 		h.Set(contentTypeHeader, mime)
 	}
+	// since transfer-encoding is being unconditionally set, content-length must not be set
```
Comment (jbenet) [inline commands/http/handler.go]: this is not a good solution. `Content-Length` is a very useful property. perhaps we should define cases where `Transfer-Encoding` is not set? 

cc @mappum thoughts here?

also @WeMeetAgain does this _actually break_ something? or it just a recommendation?

**Comment by WeMeetAgain** [comment]
It was breaking a non-conforming http client I was using, not breaking anything anymore. I'd be happy to leave as is, since any conformant http client should be ignoring content-length when a non-identity transfer-encoding is also present.

**Comment by jbenet** [comment]
@diasdavid saw something weird today. on https://ipfs.io homepage:

```
https://ipfs.io/styles/bootstrap/css/bootstrap.min.css Failed to load resource: net::ERR_CONTENT_LENGTH_MISMATCH
https://ipfs.io/styles/img/main/alpha-sunrise-2000x1682.png Failed to load resource: net::ERR_CONTENT_LENGTH_MISMATCH
chrome-extension://boadgeojelhgndaghljhdicfkmllpafd/cast_sender.js Failed to load resource: net::ERR_ADDRESS_UNREACHABLE
html5player-new.js:309 SVG's SMIL animations (<animate>, <set>, etc.) are deprecated and will be removed. Please use CSS animations or Web animations instead.
https://static.doubleclick.net/instream/ad_status.js Failed to load resource: net::ERR_BLOCKED_BY_CLIENT
2chrome-extension://boadgeojelhgndaghljhdicfkmllpafd/cast_sender.js Failed to load resource: net::ERR_ADDRESS_UNREACHABLE
3chrome-extension://dliochdbjfkdbacpmhlcpmleaejidimm/cast_sender.js Failed to load resource: net::ERR_ADDRESS_UNREACHABLE
3chrome-extension://hfaagokkkhdbgiakmmlclaapfelnkoah/cast_sender.js Failed to load resource: net::ERR_ADDRESS_UNREACHABLE
3chrome-extension://fmfcbgogabcbclcofgocippekhfcmgfj/cast_sender.js Failed to load resource: net::ERR_ADDRESS_UNREACHABLE
3chrome-extension://enhhojjnijigcajfphajepfemndkmdlo/cast_sender.js Failed to load resource: net::ERR_ADDRESS_UNREACHABLE
3chrome-extension://eojlgccfgnjlphjnlopmadngcgmmdgpk/cast_sender.js Failed to load resource: net::ERR_ADDRESS_UNREACHABLE
Chrome - Version 45.0.2454.101 
```
but then it stopped. may be worth merging to avoid this.

though i wonder what this was..... how would the length vary?

**Comment by whyrusleeping** [comment]
We had decided in another PR that we should set the header `X-Content-Length` or similar, so clients who want the info have it, but so that we also dont break things that make poor assumptions.

**Comment by jbenet** [comment]
We should also always return correct length-- have seen it off sometimes

**Comment by whyrusleeping** [comment]
@WeMeetAgain @jbenet update here?

**Comment by WeMeetAgain** [comment]
Not sure about the _incorrect_ content length being sent, but I'll shortly update the PR to return the `X-Content-Length` header instead of simply suppressing the content length.

**Comment by WeMeetAgain** [comment]
updated~, OP edited to reflect changes.

**Comment by jbenet** [comment]
maybe this is a good way to avoid breaking things if the thing is incorrect header.

before merging, feedback @mappum @diasdavid ?

**Comment by diasdavid** [comment]
It has been a long time since I saw that error. Setting up Content-Length LGTM as it is the standardised header for that purpose

**Comment by jbenet** [comment]
@diasdavid this is not using `Content-Length` but rather _making sure we don't use it_. @mappum thoughts?

**Comment by RichardLitt** [comment]
I think switching to `X-Content-Length` is a smart move. But I'm not sure we need to provide `content-length` at all for this; `cat` shouldn't be returning it .

**Comment by lgierth** [comment]
LGTM

**Comment by whyrusleeping** [comment]
I believe this has been resolved in a different PR, i no longer see both content length and chunked encoding together. If it is still an issue please reopen this PR, or file an issue with repro steps

**Pull request closed without merging**